### PR TITLE
[#1698] Refactoring of translations

### LINKED ISSUE #1698: Refactoring of translations (whitelabel solution)
## Overview

Move application's translation from default `i18n` to gettext ([Fast Gettext](https://github.com/grosser/fast_gettext)) or equivalent.

The idea is to have all content in default language (english) directly in views (hamls).

## Specifics

Note - Fast Gettext might not be the best current solution - further research is required as part of this issue's scope.
What is required is a `gettext`ish solution which will make translations as transparent as possible for developers

## Scope

All application id based translation's in views / code (the ones in code should be non existent or very rare) should be changed to gettext translations.

Gettext string extraction should be run to validate whether the solution works as expected.

Things **not included** in the scope of this issue

* Gem specific translations (simple form / devise) should be left as is.
* "Tranlations" / Yaml files which are part of the 'tours' feature should be left as is, those are not translations in the strict sense of the word, and will be included in feature refactoring works.

## Possible issues / things to keep in mind

There are some issues of dynamic id generations for translation purposes:
`"title": "#{t("project_items.order_type.#{order_type(service)}.tooltip")}"`

It is advised to note all occurrences of such translation usage to make further refactoring easier in the future.

I would advise to keep them as is for now, or coordinate with #1702 in order to avoid conflicts and clashes.
Translation / extraction of strings related to ordering can be done as part of this issue in separate PR.

### PR DESCRIPTION
**Pull request by JanKapala**
closes #1698

Description of the PR is available in the issue that it closes. Below is a list of sub tasks:

### TODO:
- [x] best gem research: there are two interesting choices: fast_gettext i gettext. First is much better:
       - it’s order of magnitude faster
       - it uses two orders of magnitude less memory
- [x] views refactoring
  - [x] static translations
  - [x] part of dynamic translations
  - [x] html insertions
  -  [x] convert direct strings to translations
  - [x] mark missing translations with `-# TODO: missing translation (no entry in original i18 locales)`
  - [x] mark dynamic translations (actually constant -> string mappings that are not exactly translations) with `-# TODO: refactor dynamic translation`
  - [x] translations of links (`link_to` and hardcoded hrefs (except footer partial - see below))
- [x] configuration of POs

### PR Evaluation method (for reviewers):
In the browser check text on all pages that are associated with modified views (time consuming, I know :( )

### Next related PR work directions:
- above changes were made only in the `marketplace/app/views directory, but there are also strings in e.g. controllers etc. - both direct string and old i18 translations
- refactor remaining dynamic translations (actually constant -> string mappings that shouldn't be done with translations) using for example rails helpers
- files that probably should be deleted:
    - `marketplace/app/views/designsystem/index.html.haml`
    - `marketplace/app/views/layouts/designsystem.html.haml`
- not dry file (loop needed): `marketplace/app/views/services/_details.html.haml`
- missing translations marked in following files:
    - `marketplace/app/views/backoffice/platforms/show.html.haml`
    - `marketplace/app/views/designsystem/index.html.haml`
    - `marketplace/app/views/projects/index.html.haml`
- translations of mailer:
    - `marketplace/app/views/layouts/mailer.text.haml`
    - `marketplace/app/views/project_item_mailer/*`
    - `marketplace/app/views/service_mailer/new_question.html.haml`
    - `marketplace/app/views/webhook_jira_mailer/message_edited.html.haml`
- removing all .yaml locales (using PO files if needed)
- replace html hrefs with rails link_to in the footer partial: `marketplace/app/views/layouts/_footer.html.haml`

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -41,4 +42,11 @@ def not_authorized_message(exception)
       policy_name = exception.policy.class.to_s.underscore
       I18n.t "#{policy_name}.#{exception.query}", scope: :pundit, default: :default
     end
+
+    def set_locale
+      FastGettext.available_locales = ["en"]
+      FastGettext.text_domain = "marketplace"
+      FastGettext.set_locale(params[:locale] || session[:locale] || request.env["HTTP_ACCEPT_LANGUAGE"])
+      session[:locale] = I18n.locale = FastGettext.locale
```
Comment (michal-szostak) [inline app/controllers/application_controller.rb:50]: We do not really support translations, so we should hard code `en_GB` here

File: locale/README.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+# Folder `locale` is needed for fast_gettext gem and git doesn't commit empty directories that's why this file exist.
```
Comment (michal-szostak) [inline locale/README.txt:1]: In these cases the standard procedure is to add empty file called `.gitkeep` or something like that

**Review by michal-szostak — CHANGES_REQUESTED** [review]
I'm happy that it generally works. There are a few things to fix.

Also: 

1. When creating PR the convention is to name the PR as `[#<issue_id>] <Title>
2. In PR it's good to describe the scope, including screenshots is also a nice pracrice, if issue touches UI - it helps in finding feature in the application itself. In case of translations that last part is not required though.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (JanKapala) [reply app/controllers/application_controller.rb:50]: en_GB or just en?

File: locale/README.txt (hunk 2 above)
Comment (JanKapala) [reply locale/README.txt:1]: good to know! :)

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (JanKapala) [reply app/controllers/application_controller.rb:50]: Ad 2; I could rewrite a description of issue closed by this PR as a description of this PR but I wonder if is it necessary

**Comment by agpul** [comment]
Now, I haven't found any problems from my point of view. I will make a final review at the end of the implementation.

File: app/views/projects/_project_layout.html.haml [hunk 3]
```diff
@@ -6,6 +6,7 @@
           %h1.flex-grow-1.text-break
             = project.name
           .details-row
+            -# TODO: refactor dynamic translation
             - if project.archived?
               .details-box-info.text-uppercase= t("project_items.status.#{project.status}")
```
Comment (wujuu) [inline app/views/projects/_project_layout.html.haml:11]: Translation missing

File: app/views/projects/edit.html.haml [hunk 4]
```diff
@@ -1,16 +1,17 @@
-- content_for :title, "#{t(".title")} #{@project.name}"
+-# TODO: refactor dynamic translation
```
Comment (wujuu) [inline app/views/projects/edit.html.haml:1]: There already is a dynamic tranlastion here?

File: app/views/projects/index.html.haml [hunk 5]
```diff
@@ -1,6 +1,7 @@
-- content_for :title, t(".title")
+- content_for :title, _("My projects")
 - breadcrumb :projects
 
 = render "projects/layout", project: @project, projects: @projects do
   .empty-project.container.shadow-sm
+    -# TODO: missing translation (no entry in original i18 locales)
```
Comment (wujuu) [inline app/views/projects/index.html.haml:6]: Reminder

File: app/views/projects/services/_details.html.haml [hunk 6]
```diff
@@ -1,26 +1,33 @@
...
   %dd= link_to project_item.service.name, project_item.service
 - if project_item.service.offers_count > 1
   %dl
-    %dt Resource offer:
+    %dt
+      = _("Resource offer") + ":"
     %dd= project_item.name
 %dl
+  -# TODO: refactor dynamic translation
   %dt= t("project_items.service.label.#{map_view_to_order_type(project_item)}")
```
Comment (wujuu) [inline app/views/projects/services/_details.html.haml:12]: Dynamic translations missing

File: app/views/projects/services/_layout.html.haml [hunk 7]
```diff
@@ -10,26 +10,28 @@
...
+              = link_to _("back to %{pn} project services") % { pn: project.name }, [project, :services]
       .col-md-4.float-right.justify-content-center
 
         - if project_item.orderable?
           %p.text-uppercase.text-right.font-weight-bold.text-secondary.mb-0
+            -# TODO: refactor dynamic translation
             %span{ class: "status-#{project_item.status}" }= t("project_items.status.#{project_item.status}")
 
         - elsif webpage(project_item).present?
           = link_to t("common.#{map_view_to_order_type(project_item)}.link"), webpage(project_item),
```
Comment (wujuu) [inline app/views/projects/services/_layout.html.haml:22]: Translation missing

File: app/views/projects/services/_project_item.html.haml [hunk 8]
```diff
@@ -2,8 +2,9 @@
   = link_to project_item.service.name, project_service_path(project, project_item)
   .float-right.status
     - if project_item.closed?
+      -# TODO: refactor dynamic translation
       = t("project_items.status.#{project_item.status}")
```
Comment (wujuu) [inline app/views/projects/services/_project_item.html.haml:6]: Translation missing

File: Gemfile [hunk 9]
```diff
@@ -46,7 +49,6 @@ gem "omniauth_openid_connect"
 gem "pundit", "~> 2.0"
 gem "role_model"
 gem "recaptcha", require: "recaptcha/rails"
-gem "split", require: "split/dashboard"
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply Gemfile:49]: Look at line 70

File: app/views/designsystem/index.html.haml [hunk 10]
```diff
@@ -1,3 +1,4 @@
+-# TODO: missing translation (no entry in original i18 locales)
 - content_for :title, t(".title")
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply app/views/designsystem/index.html.haml:2]: The translation is missing as you can see in the comment ;) Also designsystem directory is probably to remove - out of the scope of this PR

File: app/views/executive/statistics/index.html.haml [hunk 11]
```diff
@@ -1,31 +1,39 @@
...
+  -# haml-lint:enable MultilinePipe
+  = render "statistic", title: _("All resources"), amount: @all_services_count,
+    description: _("All resources registered in Marketplace")
 
 #statistics.accordion
   = render "pages/card", parent: "statistics",
-    name: "providers", title: t(".providers.title", count: @providers_count) do
-    %p= t("executive.statistics.index.providers.description")
+    name: "providers", title: _("Providers (%{count})") % { count: @providers_count } do
+    -# TODO: refactor dynamic translation (above)
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply app/views/executive/statistics/index.html.haml:20]: this is obsolete TODO  comment, alreade refactored

File: app/views/home/communities/_card.html.haml [hunk 12]
```diff
@@ -1,3 +1,4 @@
+-# TODO: refactor dynamic translation
 .title= t(".#{key}.title")
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply app/views/home/communities/_card.html.haml:2]: this is dynamic translation or rather using translation to do something else than translations: maping constant to strings - so it's out of the PR scope

File: app/views/projects/edit.html.haml (hunk 4 above)
Comment (JanKapala) [reply app/views/projects/edit.html.haml:1]: It's obsolete TODO comment as well

File: app/views/projects/index.html.haml (hunk 5 above)
Comment (JanKapala) [reply app/views/projects/index.html.haml:6]: Yeah, for now we keep missing translations as they are. Same as before

File: app/views/projects/services/_details.html.haml (hunk 6 above)
Comment (JanKapala) [reply app/views/projects/services/_details.html.haml:12]: As before: dynamic translations should be marked only, no refactoring in this PR

File: app/views/projects/services/_layout.html.haml (hunk 7 above)
Comment (JanKapala) [reply app/views/projects/services/_layout.html.haml:22]: As before: dynamic translations should be marked only, no refactoring in this PR

File: app/views/projects/services/_project_item.html.haml (hunk 8 above)
Comment (JanKapala) [reply app/views/projects/services/_project_item.html.haml:6]: As before: dynamic translations should be marked only, no refactoring in this PR

File: app/views/executive/statistics/index.html.haml (hunk 11 above)
[the remark this answers is not in the document]
Comment (wujuu) [reply app/views/executive/statistics/index.html.haml]: So can't we delete this comment?

File: app/views/projects/edit.html.haml (hunk 4 above)
Comment (wujuu) [reply app/views/projects/edit.html.haml]: Well then it shouldn't be here

File: config/initializers/fast_gettext.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+
+# config/initializers/fast_gettext.rb
+FastGettext.add_text_domain "marketplace", path: "locale", type: :po
+FastGettext.default_available_locales = ["en", "pl"]
+FastGettext.default_text_domain = "marketplace"
```
Comment (wujuu) [inline config/initializers/fast_gettext.rb:5]: No newline

File: lib/templates/haml/scaffold/_form.html.haml [hunk 14]
```diff
@@ -1,12 +1,12 @@
...
+-#   = f.error_notification
+-#   = f.error_notification message: f.object.errors[:base].to_sentence if f.object.errors[:base].present?
+-#
+-#   .form-inputs
+-#   <%- attributes.each do |attribute| -%>
+-#     = f.<%= attribute.reference? ? :association : :input %> :<%= attribute.name %>
+-#   <%- end -%>
+-#
+-#   .form-actions
+-#     = f.button :submit
```
Comment (wujuu) [inline lib/templates/haml/scaffold/_form.html.haml:12]: We don't need simple form ????

Comment (JanKapala) [reply lib/templates/haml/scaffold/_form.html.haml:12]: I don't know, maybe, but it is in plain erb - not haml - and it was the cause of gettext parser failure :/

**Review by agpul — CHANGES_REQUESTED** [review]
Please check all places one more time, I've tried to catch all problems but I think that we need a double check.
1. The text on the main page is in polish (should be in english)
image
2. Title? and More HTML?
image

3.  Project page doesn't work! - I am not able to check pages related to the project concept. 
image
https://marketplace-2.docker-fid.grid.cyf-kr.edu.pl/projects
4. Browse service page:
image
5. When the mouse is on the link I see something strange:
image
6. Wrong  button name:
image

**Comment by JanKapala** [comment]
@agpul I couldn't reproduce these errors locally, problems doesn't occur apart from polish description at home - I've fix it.
Maybe have you used wrong testing instance?

**Review by agpul — APPROVED** [review]

**Review by michal-szostak — APPROVED** [review]

**Review by michal-szostak — APPROVED** [review]

**Review by wujuu — APPROVED** [review]

**Pull request merged**